lib/puma/request.rb

```diff
@@ -196,25 +185,74 @@ def default_server_port(env)
     #
     def fast_write(io, str)
       n = 0
-      while true
+      byte_size = str.bytesize
+      while n < byte_size
         begin
-          n = io.syswrite str
-        rescue Errno::EAGAIN, Errno::EWOULDBLOCK
-          if !IO.select(nil, [io], nil, WRITE_TIMEOUT)
+          n += io.write_nonblock(n == 0 ? str : str.byteslice(n..-1))
+        rescue IO::WaitWritable, Errno::EINTR
+          unless IO.select(nil, [io], nil, WRITE_TIMEOUT)
             raise ConnectionError, "Socket timeout writing data"
           end
-
           retry
         rescue  Errno::EPIPE, SystemCallError, IOError
           raise ConnectionError, "Socket timeout writing data"
         end
-
-        return if n == str.bytesize
-        str = str.byteslice(n..-1)
       end
     end
     private :fast_write
 
+    def __fast_write(io, str)
+      io.write_nonblock str
+    rescue IO::WaitWritable, Errno::EAGAIN, Errno::EWOULDBLOCK
+      if !IO.select(nil, [io], nil, WRITE_TIMEOUT)
+        raise ConnectionError, "Socket timeout writing data"
+      end
+      retry
+    rescue Errno::EPIPE, SystemCallError, IOError
+      raise ConnectionError, "Socket timeout writing data"
+    end
+
+    # Writes an array to an io (normally Client#io) using #syswrite.
+    # @param io [#syswrite] the io to write to
+    # @param enum [Enumerable] the body object
+    # @raise [ConnectionError]
+    #
+    def fast_write_ary(io, enum, strm)
+      running_len = 0
+      enum.each do |part|
+        next if (byte_size = part.bytesize).zero?
+        running_len += byte_size
+        if running_len > BUFFER_LENGTH && byte_size != running_len
+          io.write_nonblock strm.read
```

Comment:
Needs to handle `IO::WaitWritable` and `Errno::EINTR`, also write_nonblock may perform partial write according to the documentation.

Reply:
Thanks.  That was a mistake.  Fixed in push just now.  And you're correct, `write_nonblock` & `syswrite` both may do partial writes for large strings.